User can add a poem

whyderrick opened the pull request:
This first PR sets up what I expect to be the core model of the app, currently `Poem` but will probably be abstracted out to something more generic in the future. As much as possible, this PR tried to run by pure TDD.

Commit:
Allow User to add a Poem

This commit creates the first core function of this application: allow a
user to register and subsequently create a poem that's saved to the
application's database.

Most of what's contained in this commit is straightforward. User auth is
implemented by [Clearance](https://github.com/thoughtbot/clearance) and
the logic for poems is pretty anemic at this point. The values for
`Poem#status` are preliminary at this point, but their purpose is to
track a poem though the submission process.

The biggest thing worth explaining here is the removal of Skylight from
the application's Gemfile, relative to what
[Suspenders](https://github.com/thoughtbot/suspenders) output. It was
interfering with tests running and the project doesn't currently need a
monitoring solution. If it does Skylight should be re-added and
configured.

gylaz commented on app/controllers/poems_controller.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+
+  def create
+    @poem = Poem.new(poem_params)
```
Could call `Poem.create` here and remove the call to `save`.

whyderrick replied:
👍

gylaz commented on app/views/home/index.html.erb:
```diff
@@ -0,0 +1,18 @@
+    <% else %>
+      <p>Looks like you don't have any poems yet.</p>
+    <% end %>
+    <%= button_to t("poems.actions.new"), new_poem_path, method: :get %>
```
Any reason you used a `button_to` here? Is it for the visual look? Functionally, a link would be a better semantic HTML element to use. Consider using a `link_to` but using CSS to style it to look like a button.

whyderrick replied:
I was trying to focus on wiring up the functionality and not worry about applying any css classes until I had wireframes together. Certainly make sense to stick with `link_to` for `get` and `button_to` for `post` instead of overriding the method.

gylaz commented on app/views/layouts/application.html.erb:
```diff
@@ -14,6 +14,7 @@
   <%= csrf_meta_tags %>
 </head>
 <body class="<%= body_class %>">
+  <%= render "session_bar" -%>
```
I believe the `-` no longer has any effect.

whyderrick replied:
good to know! I looked up what it did when Responders prompted it, but didn't make sure it hadn't been deprecated.

gylaz commented on app/views/poems/_form.html.erb:
```diff
@@ -0,0 +1,13 @@
+<%= simple_form_for (@poem) do |form| %>
```
Remove the space before the `(`?

whyderrick replied:
👍

gylaz commented on app/views/poems/_form.html.erb:
```diff
@@ -0,0 +1,13 @@
+      
+  <%= form.input :title %>
+  <%= form.select :status,
+        ["Available", "Pending", "Accepted – Forthcoming", "Published"]
```
Worth moving these dropdown options to a helper method?

whyderrick replied:
Absolutely worth it. Thanks for the suggestion

gylaz commented on app/views/poems/_form.html.erb:
```diff
@@ -0,0 +1,13 @@
+  %>
+  <%= form.input :user_id, 
+        as: :hidden,
+        input_html: { value: current_user.id }
```
Typically, we indent one extra level in regards to the previous line, which includes the `<%=`.

whyderrick replied:
👍

gylaz commented on db/migrate/20170414211503_create_users.rb:
```diff
@@ -0,0 +1,14 @@
+      t.string :remember_token, limit: 128, null: false
+    end
+
+    add_index :users, :email
```
Worth making this index be unique?

whyderrick replied:
definitely worthwhile. I left it as it came from `Clearance`

gylaz commented on db/migrate/20170414233058_create_poems.rb:
```diff
@@ -0,0 +1,13 @@
+
+  def down
+    drop_table :poems
+  end
```
Since your `up` is adding this table, you can get away with using `change`.

whyderrick replied:
sounds good. I wasn't sure if we just steered clear of `change` in favor for the added control.

gylaz commented on db/schema.rb:
```diff
@@ -30,4 +30,26 @@
+  end
+
+  create_table "poems_tables", force: :cascade do |t|
+  end
```
Where did this table come from?

whyderrick replied:
Typo. I deleted the migration without rolling back the migration I think.

gylaz commented on spec/features/poem_spec.rb:
```diff
@@ -0,0 +1,46 @@
+
+      click_on t('poems.actions.new')
+      fill_in :poem_title, with: poem_title
+      select "Pending", from: :poem_status
```
How about using `poem_status` here?

whyderrick replied:
👍

gylaz commented on spec/features/poem_spec.rb:
```diff
@@ -0,0 +1,46 @@
+      # be to realize that the page blew up here in the client.
+
+      expect(page).to have_flash_message(:notice, text: "Poem was created successfully")
+      expect(page).to have_content(poem_title)
```
I recommend using `have_text` instead of `have_content`.

`have_content` is an alias of `have_text`.

whyderrick replied:
Thanks. Useful.

gylaz commented on spec/features/poem_spec.rb:
```diff
@@ -0,0 +1,46 @@
+      poem_status = "Pending"
+
+      visit root_path(as: user)
+
```
Does this newline serve a purpose?

whyderrick replied:
Nope. I didn't commit to grouping it with the setup or the execution.

gylaz commented on spec/models/poem_spec.rb:
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
```
Newline after `require`?

whyderrick replied:
👍

gylaz commented on spec/models/poem_spec.rb:
```diff
@@ -0,0 +1,10 @@
+  describe "associations" do
+    it { is_expected.to belong_to(:user) }
+  end
+  describe "validations" do
```
Newline between `describe` blocks?

whyderrick replied:
👍

gylaz commented on spec/support/matchers/flash_matchers.rb:
```diff
@@ -0,0 +1,11 @@
+  match do |page|
+    if flash_type.present?
+      page.has_css?("div.flash_#{flash_type}")
+      within("div.flashes") { page.has_text }
```
Does `has_text`get passed something? Or does it assert that *any* text is present?

Can you achieve something similar in on shot via `page.has_css?("div.flash_#{flash_type}", text: options)`?

whyderrick replied:
Thanks. That make sense and will definitely help me make better use of Capybara's finders.

whyderrick committed:
Feedback commit

This commit adds a handful of small changes based on PR feedback.
- Semantic use of `link_to` in home/index.html.erb required adding a
`button` class to _buttons.scss.
- Using responders for a failed update required making the `alert`
messages availalbe in its translation file.

whyderrick commented:
![](https://github.trello.services/images/mini-trello-icon.png) [As a user, I can create a poem](https://trello.com/c/VK0DGAmB/9-as-a-user-i-can-create-a-poem)
